src/edu/csus/ecs/pc2/ui/AutoJudgeStatusFrame.java

```diff
@@ -138,10 +143,14 @@ private JPanel getMessagePanel() {
     private JButton getStopAutoJudgingButton() {
         if (stopAutoJudgingButton == null) {
             stopAutoJudgingButton = new JButton();
-            stopAutoJudgingButton.setText("Stop AutoJudging");
+            stopAutoJudgingButton.setText("Stop Auto Judging");
             stopAutoJudgingButton.addActionListener(new java.awt.event.ActionListener() {
                 public void actionPerformed(java.awt.event.ActionEvent e) {
-                    startStopAutoJudging();
+                    if(notifyFrame != null) {
+                        notifyFrame.executeFrameTerminated();
+                    } else {
+                        startStopAutoJudging();
+                    }
```

Comment:
I'm not at all clear about the changes in the above `actionPerformed()` method.  Previously, pressing the "start/stop AutoJudging" button (i.e., invoking this method) would unconditionally invoke method `startStopAutoJudging()`, which serves two functions: (1) toggling the label on the "Start/Stop AutoJudging" button between Stop/Start, and (2) enabling or disabling the `AutoJudgingMonitor` (that is, making the AJ start/stop auto-judging runs as appropriate).

Now, if `notifyFrame != null`, this method is never called.  This doesn't seem logical to me.  It appears to me that the intent of the PR is to have the **Terminate** button notify the `notifyFrame` that the operator (user) has terminated the current execution.  That seems to me to be something totally unrelated to what happens when the `startStopAutoJudging` button is pressed; the former has to do with terminating a Submission execution while the latter has to do with enabling/disabling the AJ's ability to fetch/judge runs, which are two totally separate things.

Or am I totally misunderstanding the structure of the PR?  (Which is possible; it's quite complex and I'm struggling to get a clear understanding of what's going on...)

Reply:
The handling of the `Terminate `/ `Start-Stop AJ `button is done in one place now.  If `notifyFrame `is non-null, it means that a run is taking place (the  execute timer is running), in which case, the button should only terminate the judging currently going on and not stop _Autojudging_ altogether (IE the `Terminate `button).  If `notifyFrame `is null, it means it is either between test cases or it has finished judging the last submission and there are no more submissions waiting (IE Stop Autojuding).  This all has to do with the fact that there is a single frame now, rather than a new (separate one) for each test case as well as one for the quiescent case (judge queue empty).   Does that make sense?

Reply:
You said:
> The handling of the `Terminate `/ `Start-Stop AJ `button is done in one place now. If `notifyFrame `is non-null, it means that a run is taking place (the execute timer is running), in which case, the button should only terminate the judging currently going on and not stop _Autojudging_ altogether (IE the `Terminate `button). If `notifyFrame `is null, it means it is either between test cases or it has finished judging the last submission and there are no more submissions waiting (IE Stop Autojuding). This all has to do with the fact that there is a single frame now, rather than a new (separate one) for each test case as well as one for the quiescent case (judge queue empty). Does that make sense?

Not yet, although I grant that it's possible it's my lack of understanding.  But here's what I see:
  
You said: `The handling of the Terminate / Start-Stop AJ button is done in one place now.` That may be, but those are _not_ the same thing.  The `Terminate` button is used by a (human) judge to terminate the execution of a test case which is taking too long (but the human judge doesn't want to just let it run until the runtime-limit is reached).  The `Start-Stop AJ` button is used to enable/disable the ability of a Judge client to automatically fetch/execute/validate runs.  Those are two different operations, and it's not clear to me how (or more correctly, why) the new code is trying to make them work the same (which is what it appears to me that it's trying to do).

Also, you said `and there are no more submissions waiting (IE Stop Autojuding)`.  Those are two separate (and relatively unrelated) conditions (that is, "`there are no more submissions waiting`" is a different condition from" `Stop AutoJudging`").  When "there are no more submissions waiting", the AJ simply waits, polling on occasion to find out if a new run has been submitted.  That is _not_ the same condition as "Stop AutoJudging"; the latter means "turn off the judge client's ability to poll for and to pull runs to be judged (i.e., switch to "manual, non-auto" judging mode).  So I'm still confused about what the code is trying to do.

Reply:
Maybe I didn't explain it clearly.  Think of the `notifyFrame `member as an indication that a run is being executed at the moment if it is non-null.   In that case, the semantics of the `Terminate `button correspond to those of terminating a running submission.  If `notifyFrame `is null, it means that the AJ is not currently executing a submission, rather, it is sitting there waiting for more runs;  the user can now press the `Stop Autojudging` button.  Basically, the button is being "doubled-up" on depending on what is going on.  If executing a submission, it terminates the run.  If waiting for a submission, it stops autojudging.

`notifyFrame `is set by a call to `setTerminateButtonNotify(...)` from `ExecuteTimer `when a submission starts executing.  
It is set to **null** when the `ExecuteTimer `finishes (`ExecuteTimer.stopTimer(...)`) (at which point the `AutoJudgeStatusFrame `is again, waiting for new runs).

We can always have a google meet and trace through it if that helps.